src/transformers/models/umt5/modeling_flax_umt5.py

```diff
@@ -0,0 +1,1794 @@
+        key_states = self.k(hidden_states) if key_value_states is None else self.k(key_value_states)
+        value_states = self.v(hidden_states) if key_value_states is None else self.v(key_value_states)
+
+        # reshape to (batch_size, seq_length, n_heads, head_dim)
+        query_states = self._split_heads(query_states)
+        key_states = self._split_heads(key_states)
+        value_states = self._split_heads(value_states)
+
+        # counter-act scaling in dot_product_attention_weights function
+        query_states *= jnp.sqrt(query_states.shape[-1])
+
+        # for fast decoding causal attention mask should be shifted
+        causal_attention_mask_shift = (
+            self.variables["cache"]["cache_index"] if (self.has_variable("cache", "cached_key") and self.causal) else 0
+        )
+        # create causal attention_mask; attention_mask has to be defined when model is causal
+        if self.causal:
+            causal_attention_mask = make_causal_mask(attention_mask, dtype="bool")
+
+            # fast decoding for generate requires special attention_mask
+            if self.has_variable("cache", "cached_key"):
+                max_decoder_length = self.variables["cache"]["cached_key"].shape[1]
+                causal_attention_mask = jax.lax.dynamic_slice(
+                    causal_attention_mask,
+                    (0, 0, causal_attention_mask_shift, 0),
+                    (1, 1, seq_length, max_decoder_length),
+                )
+
+            # broadcast causal attention mask & attention mask to fit for merge
+            causal_attention_mask = jnp.broadcast_to(
+                causal_attention_mask, (batch_size,) + causal_attention_mask.shape[1:]
+            )
+            attention_mask = jnp.broadcast_to(
+                jnp.expand_dims(attention_mask, axis=(-3, -2)), causal_attention_mask.shape
+            )
+            attention_mask = combine_masks(attention_mask, causal_attention_mask)
+        elif attention_mask is not None:
+            attention_mask = jnp.expand_dims(attention_mask, axis=(-3, -2))
+
+        # During fast autoregressive decoding, we feed one position at a time,
+        # and cache the keys and values step by step.
+        if self.causal and (self.has_variable("cache", "cached_key") or init_cache):
+            key_states, value_states, attention_attention_mask = self._concatenate_to_cache(
+                key_states, value_states, query_states, attention_mask
+            )
+
+        # replace masked positions with -10_000
+        if attention_mask is not None:
+            mask_value = jnp.finfo(self.dtype).min
+            attention_mask = jax.lax.select(
+                attention_mask > 0,
+                jnp.full(attention_mask.shape, 0.0).astype(self.dtype),
+                jnp.full(attention_mask.shape, mask_value).astype(self.dtype),
+            )
+
+        if position_bias is None:
```

Comment:
When I remove this check (and fix identation), then I can see the `relative_attention_bias` key in all layers:

```
Layer: 0
dict_keys(['k', 'o', 'q', 'relative_attention_bias', 'v'])
Layer: 1
dict_keys(['k', 'o', 'q', 'relative_attention_bias', 'v'])
Layer: 2
dict_keys(['k', 'o', 'q', 'relative_attention_bias', 'v'])
Layer: 3
dict_keys(['k', 'o', 'q', 'relative_attention_bias', 'v'])
Layer: 4
dict_keys(['k', 'o', 'q', 'relative_attention_bias', 'v'])
Layer: 5
dict_keys(['k', 'o', 'q', 'relative_attention_bias', 'v'])
Layer: 6
dict_keys(['k', 'o', 'q', 'relative_attention_bias', 'v'])
Layer: 7
dict_keys(['k', 'o', 'q', 'relative_attention_bias', 'v'])

```

Reply:
Hmm, true, but this means more changes are needed because position_bias shot not be shared across all layers.
The same goes for pytorch, in line 1081.

Reply:
Definitely. I've also updated the conversion script -> it supports per-layer relative bias now!

https://gist.github.com/stefan-it/5d6a4ec89e7ad97181983881434cb4eb?permalink_comment_id=4528774#gistcomment-4528774

Reply:
ok, I think I fixed it on line 710 and line 713. Both of the position_bias and encoder_decoder_position_bias should not be set so they won't be shared across layers.

Still the results after the conversion is not correct.

Reply:
I have updated the testing notebook here:
https://colab.research.google.com/drive/1QrqxNdIK7ugQ3FC8tqxUqZZwP0zdvYE4?usp=sharing

Reply:
Ahhhh!! I found one Copy-Paste Bug! Sorry for that!

Output is then:

```
Out[1]: '<pad> Wikipedia is a Wikipedia Wikipedia is a Wikipedia Wikipedia is a Wikipedia Wikipedia is a Wikipedia Wikipedia is'
```

Reply:
Hmm, no worries, we are still trying to fix bugs and make it work 😄 
I have tested the new model + the latest push from my side and it is still not working correctly:
`test_model = FlaxAutoModelForSeq2SeqLM.from_pretrained("stefan-it/umt5-small")
````
WARNING:jax._src.xla_bridge:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Some weights of the model checkpoint at stefan-it/umt5-small were not used when initializing FlaxUMT5ForConditionalGeneration: {('decoder', 'block', '4', 'layer', '1', 'EncDecAttention', 'relative_attention_bias', 'embedding'), ('decoder', 'block', '2', 'layer', '1', 'EncDecAttention', 'relative_attention_bias', 'embedding'), ('decoder', 'block', '7', 'layer', '1', 'EncDecAttention', 'relative_attention_bias', 'embedding'), ('decoder', 'block', '6', 'layer', '1', 'EncDecAttention', 'relative_attention_bias', 'embedding'), ('decoder', 'block', '3', 'layer', '1', 'EncDecAttention', 'relative_attention_bias', 'embedding'), ('decoder', 'block', '5', 'layer', '1', 'EncDecAttention', 'relative_attention_bias', 'embedding'), ('decoder', 'block', '1', 'layer', '1', 'EncDecAttention', 'relative_attention_bias', 'embedding')}
- This IS expected if you are initializing FlaxUMT5ForConditionalGeneration from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing FlaxUMT5ForConditionalGeneration from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
```
At the very end, it produces garbage.

Reply:
Oh, I also saw this text and then I uploaded the `tokenizer_config.json` -> maybe you can force a clean pull from model hub for the tokenizer :thinking:

Reply:
Some good news:

I made an experiment on fine-tuning `squad_v2` with mT5 Small and umT5  Small:

| Model  | Train loss  | Eval loss
| ---------- | --------------- | ------------
| `google/mt5-small` | 1.6894 | 1.2819
| `stefan-it/umt5-small` | 1.7913 | 1.4628

For one epoch. I have no idea, why EM/F1 is no longer shown after fine-tuning (???), but yeah, there's a little performance difference. But we are on a good way!

Reply:
Great, thanks for testing it on this downstream model.

There are three variables here:
1. The tokenizer.
2. The conversion.
3. The model.

I checked the tokenizer, and it did not correctly encode the <extra_id_0>. It was simply breaking it down to multiple ids rather than a single id.
I have corrected this issue, and I now have the correct tokenizer here:
[agemagician/umt5-small](https://huggingface.co/agemagician/umt5-small/tree/main)

I have tested it carefully, and it should be correct.

This leaves the model conversion and the model script at hugging face. I will now investigate them.

Reply:
Hi @agemagician ,

ah, I see one difference is that `extra_ids` stuff. You set them to ` "extra_ids": 0` - as it is done in mT5.

But there's a vocab difference between mT5 and umT5:

mT5 has a vocab size of 250112 (in `config.json`). Then I loaded the SPM vocab:

```python
>>> import sentencepiece as spm
>>> vocab_file = "./spiece-mt5.model"
>>> sp_model = spm.SentencePieceProcessor()
>>> sp_model.Load(vocab_file)
>>> sp_model.vocab_size()
250100
```

Then I found out, that they added the `<extra_id_0>` stuff directly in the SPM vocab:

```python
>>> sp_model.IdToPiece(250099)
'▁<extra_id_0>'
>>> sp_model.IdToPiece(250098)
'▁<extra_id_1>'
>>> sp_model.IdToPiece(250000)
'▁<extra_id_99>'
>>> sp_model.IdToPiece(250000)
'▁<extra_id_99>'
```

Then I looked at the umT5. The embedding dimension is 256384. I loaded the SPM vocab for umT5:

```python
>>> import sentencepiece as spm
>>> vocab_file = "spiece-umt5.model"
>>> sp_model = spm.SentencePieceProcessor()
>>> sp_model.Load(vocab_file)
True
>>> sp_model.vocab_size()
256000
```

And tried to find the `<extra_id_*> stuff:

```python
>>> sp_model.IdToPiece(255999)
'啕'
>>> for piece in [sp_model.IdToPiece(i) for i in range(0, sp_model.vocab_size())]:
...   if "extra_id" in piece:
...     print(piece)
>>>
```

So they did not include the `<extra_id_*> stuff in the umT5 vocab.

That's why I added `"extra_ids": 384` in the `tokenizer_config.json`.

But in this example it should correctly tokenize the `<extra_id_0>`:

```
>>> from transformers import T5TokenizerFast
>>> tokenizer = T5TokenizerFast.from_pretrained("stefan-it/umt5-small")
>>> sentence = "Wikipedia is a <extra_id_0>"
>>> tokenizer.tokenize(sentence)
['▁Wikipedia', '▁is', '▁a', '▁', '<extra_id_0>']
```

and:

```
>>> ids = tokenizer(sentence, return_tensors="jax")
>>> ids
{'input_ids': Array([[ 14164,    339,    289,    273, 256383,      1]], dtype=int32), 'attention_mask': Array([[1, 1, 1, 1, 1, 1]], dtype=int32)}
```

So '<extra_id_0>' points to 256383.

:thinking:

Reply:
Oh,

but this number is not correct:

https://github.com/google-research/t5x/blob/main/t5x/examples/scalable_t5/umt5/vocab.gin#L5

It should only be 300 extra ids - will correct it!

Reply:
> Hi @agemagician ,
> 
> ah, I see one difference is that `extra_ids` stuff. You set them to ` "extra_ids": 0` - as it is done in mT5.
> 
> But there's a vocab difference between mT5 and umT5:
> 
> mT5 has a vocab size of 250112 (in `config.json`). Then I loaded the SPM vocab:
> 
> ```python
> >>> import sentencepiece as spm
> >>> vocab_file = "./spiece-mt5.model"
> >>> sp_model = spm.SentencePieceProcessor()
> >>> sp_model.Load(vocab_file)
> >>> sp_model.vocab_size()
> 250100
> ```
> 
> Then I found out, that they added the `<extra_id_0>` stuff directly in the SPM vocab:
> 
> ```python
> >>> sp_model.IdToPiece(250099)
> '▁<extra_id_0>'
> >>> sp_model.IdToPiece(250098)
> '▁<extra_id_1>'
> >>> sp_model.IdToPiece(250000)
> '▁<extra_id_99>'
> >>> sp_model.IdToPiece(250000)
> '▁<extra_id_99>'
> ```
> 
> Then I looked at the umT5. The embedding dimension is 256384. I loaded the SPM vocab for umT5:
> 
> ```python
> >>> import sentencepiece as spm
> >>> vocab_file = "spiece-umt5.model"
> >>> sp_model = spm.SentencePieceProcessor()
> >>> sp_model.Load(vocab_file)
> True
> >>> sp_model.vocab_size()
> 256000
> ```
> 
> And tried to find the `<extra_id_*>` stuff:
> 
> ```python
> >>> sp_model.IdToPiece(255999)
> '啕'
> >>> for piece in [sp_model.IdToPiece(i) for i in range(0, sp_model.vocab_size())]:
> ...   if "extra_id" in piece:
> ...     print(piece)
> >>>
> ```
> 
> So they did not include the `<extra_id_*>` stuff in the umT5 vocab.
> 
> That's why I added `"extra_ids": 384` in the `tokenizer_config.json`.
> 
> But in this example it should correctly tokenize the `<extra_id_0>`:
> 
> ```
> >>> from transformers import T5TokenizerFast
> >>> tokenizer = T5TokenizerFast.from_pretrained("stefan-it/umt5-small")
> >>> sentence = "Wikipedia is a <extra_id_0>"
> >>> tokenizer.tokenize(sentence)
> ['▁Wikipedia', '▁is', '▁a', '▁', '<extra_id_0>']
> ```
> 
> and:
> 
> ```
> >>> ids = tokenizer(sentence, return_tensors="jax")
> >>> ids
> {'input_ids': Array([[ 14164,    339,    289,    273, 256383,      1]], dtype=int32), 'attention_mask': Array([[1, 1, 1, 1, 1, 1]], dtype=int32)}
> ```
> 
> So `<extra_id_0>` points to 256383.
> 
> 🤔

In my model repo, I have added the extra ids into the spm. If you test it you will find it in my repo.
 T5TokenizerFast should not be used because it doesn't support yet byte level fallback if the tokenizer could not tokenizer a token.
That's why we have to use the slow tokenizer + original spm.
In your example it should actually gives you the following error if you are using the latest commit from transformers:
`RuntimeError: The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option which is not implemented in the fast tokenizers. In practice this means that the fast version of the tokenizer can produce unknown tokens whereas the sentencepiece version would have converted these unknown tokens into a sequence of byte tokens matching the original piece of text.`
I am not sure why you are not seeing this error.

The original tokenizer vocab size is "256000" + "300 extra tokens" + "84 empty tokens".
The empty tokens are used to pad the embedding matrix to make it multiple of 8. This is important to make the training more efficient. However, they are not used for anything else.
The "<extra_id_0>" in this case should be number 256299.

Reply:
Ok, I used slow tokenizer:

```
>>> tokenizer = T5Tokenizer.from_pretrained("stefan-it/umt5-small")
>>> special_tokens = {"additional_special_tokens": [f"<extra_id_{i}>" for i in range(300, 0, -1)]}
>>> tokenizer.add_special_tokens(special_tokens)
>>> tokenizer.decode(256299)
'<extra_id_0>'
```